### ISSUE TITLE
Segmentation Fault and Linker Error When Building with Nuitka on macOS ARM64

### ISSUE THREAD

**Issue by jason2714** [Nuitka/Nuitka#3219]
Hi,

I've successfully built my project into a single executable file using Nuitka, but I'm encountering a segmentation fault when running it:

```
Nuitka: A segmentation fault has occurred. This is highly unusual and can
have multiple reasons. Please check https://nuitka.net/info/segfault.html
for solutions.
Embeddings Shape: (5, 384)
/private/var/folders/.../multiprocessing/resource_tracker.py:123: UserWarning: resource_tracker: process died unexpectedly, relaunching. Some resources might leak.
Nuitka: A segmentation fault has occurred. This is highly unusual and can
have multiple reasons. Please check https://nuitka.net/info/segfault.html
for solutions.
```

When building with the `--debug` flag, I receive the following linker error:

```
ld: Assertion failed: (false && "compact unwind compressed function offset doesn't fit in 24 bits"), function operator(), file Layout.cpp, line 5410.
clang: error: linker command failed with exit code 1 (use -v to see invocation)
scons: *** [internal stack trace:
...
scons: building terminated because of errors.
FATAL: Failed unexpectedly in Scons C backend compilation.
Nuitka:WARNING: Complex topic! More information can be found at https://nuitka.net/info/scons-backend-failure.html
Nuitka-Reports: Compilation crash report written to file 'nuitka-crash-report.xml'.
Fail to build.
```

**Build Script:**

```bash
poetry run python -m nuitka \
    --standalone \
    --onefile \
    --lto=no \
    --static-libpython=no \
    --include-package=torch \
    --show-scons \
    run_e5_small.py
```

**Source Code (`run_e5_small.py`):**

```python
from sentence_transformers import SentenceTransformer

model = SentenceTransformer('intfloat/multilingual-e5-small', tokenizer_kwargs={'truncation': None})
model.to('mps')
input_texts = [
    'query: how much protein should a female eat',
    "hi " * 10000
]

embeddings = model.encode(input_texts, normalize_embeddings=True)

print("Embeddings Shape:", embeddings.shape)

```

**Environment:**

- **Python:** 3.12.7
- **Poetry:** 1.8.4
- **sentence-transformers:** ^3.3.1
- **OS:** macOS 14.5 (Darwin)
- **Architecture:** ARM64
- **C Compiler:** /usr/bin/clang (clang 16.0.0)
- **Nuitka:** 2.5.1

**Error Details (`nuitka-crash-report.xml`):**

```xml
<?xml version='1.0' encoding='utf8'?>
<nuitka-compilation-report nuitka_version="2.4.11" nuitka_commercial_version="not installed" completion="error exit message (1)" exit_message="Failed unexpectedly in Scons C backend compilation.">
  <scons_error_reports>
    <scons_error_report>
      <command>"/usr/bin/clang" -o "/workspace/run_e5_small.dist/embedding_server.bin" --target=arm64-macos14.5 -g -headerpad_max_install_names -Wl,-sectcreate,constants,constants,./__constants.bin @"./@link_input.txt" -L/workspace/.pyenv/versions/3.12.7/lib -lm -lpython3.12</command>
      <stderr>
0  0x1043e047c  __assert_rtn + 72
1  0x1043705c8  ___Z13dispatchApplyIRKmZ15dispatchForEach...
...
ld: Assertion failed: (false &amp;&amp; "compact unwind compressed function offset doesn't fit in 24 bits"), function operator(), file Layout.cpp, line 5410.
clang: error: linker command failed with exit code 1 (use -v to see invocation)
      </stderr>
    </scons_error_report>
  </scons_error_reports>
</nuitka-compilation-report>
```

**Steps to Reproduce:**

1. Save the provided `run_e5_small.py` script.
2. Execute the build script using Nuitka.
3. Run the resulting executable to observe the segmentation fault.
4. Rebuild with the `--debug` flag to see the linker error.

**Possible Cause:**

It appears to be an issue with the linker (`ld`) on macOS ARM64, possibly related to handling large code sections or the way Nuitka compiles the code.

**Questions:**

- Is this a known issue with Nuitka on macOS ARM64?
- Are there any workarounds or additional flags that can be used to avoid this error?

**Comment by jason2714** [comment]
I found that the segmentation fault doesn't occur when I use the model without `sentence_transformers`. Below is the huggingface code that works without causing the issue:

```python
import torch.nn.functional as F
from torch import Tensor
from transformers import AutoModel, AutoTokenizer

def average_pool(last_hidden_states: Tensor, attention_mask: Tensor) -> Tensor:
    last_hidden = last_hidden_states.masked_fill(~attention_mask[..., None].bool(), 0.0)
    return last_hidden.sum(dim=1) / attention_mask.sum(dim=1)[..., None]

# Each input text should start with "query: " or "passage: ", even for non-English texts.
input_texts = [
    'query: how much protein should a female eat',
    'query: 南瓜的家常做法',
    "passage: As a general guideline, the CDC's average requirement of protein for women ages 19 to 70 is 46 grams per day...",
    "passage: 1.清炒南瓜丝 原料:嫩南瓜半个 调料:葱、盐、白糖、鸡精 做法: 1、南瓜用刀薄薄的削去表面一层皮..."
]

tokenizer = AutoTokenizer.from_pretrained('intfloat/multilingual-e5-small')
model = AutoModel.from_pretrained('intfloat/multilingual-e5-small')

# Tokenize the input texts
batch_dict = tokenizer(input_texts, max_length=512, padding=True, truncation=True, return_tensors='pt')
outputs = model(**batch_dict)

# Average pool the token embeddings
embeddings = average_pool(outputs.last_hidden_state, batch_dict['attention_mask'])

# Normalize embeddings
embeddings = F.normalize(embeddings, p=2, dim=1)
print("Embeddings Shape:", embeddings.shape)

# Compute similarity scores
scores = (embeddings[:2] @ embeddings[2:].T) * 100
print(scores.tolist())
```

It seems that the segmentation fault is specifically related to the usage of `sentence_transformers`. Is this something related to the way `sentence_transformers` interacts with Nuitka or macOS?

**Comment by kayhayen** [comment]
There were fixes for segfault causing regressions in 2.5.2 and 2.5.3, that is probably what you are facing.

**Comment by jason2714** [comment]
I think the problem still exists with nuitka 2.5.3
```
<nuitka-compilation-report nuitka_version="2.5.3" nuitka_commercial_version="not installed" completion="error exit message (1)" exit_message="Failed unexpectedly in Scons C backend compilation.">
  <scons_error_reports>
    <scons_error_report>
      <command>"/usr/bin/clang" -o "/workspace/run_e5_small.dist/embedding_server.bin" --target=arm64-macos14.5 -g -headerpad_max_install_names -Wl,-sectcreate,constants,constants,./__constants.bin @"./@link_input.txt" -L/workspace/.pyenv/versions/3.12.7/lib -lm -lpython3.12</command>
      <stderr>
0  0x10502847c  __assert_rtn + 721  0x104fb85c8  ___Z13dispatchApplyIRKmZ15dispatchForEachIKN6mach_o13CompactUnwind11Diff24FixupEZNK2ld17LayoutLinkedImage18writeCompactUnwindENSt3__14spanIhLm18446744073709551615EEEyRKNS7_13SectionLayoutEE3$_1EvNSA_IT_Lm18446744073709551615EEEmT0_EUlmE_EvOSG_SI__block_invoke + 02  0x197c32428  _dispatch_client_callout2 + 203  0x197c46850  _dispatch_apply_invoke3 + 3364  0x197c323e8  _dispatch_client_callout + 205  0x197c33c68  _dispatch_once_callout + 326  0x197c458a4  _dispatch_apply_invoke + 2527  0x197c323e8  _dispatch_client_callout + 208  0x197c44080  _dispatch_root_queue_drain + 8649  0x197c446b8  _dispatch_worker_thread2 + 15610  0x197ddefd0  _pthread_wqthread + 228ld: Assertion failed: (false &amp;&amp; "compact unwind compressed function offset doesn't fit in 24 bits"), function operator(), file Layout.cpp, line 5410.clang: error: linker command failed with exit code 1 (use -v to see invocation)
</stderr>
```
